### PR TITLE
Reject deployment with invalid workflows

### LINKED ISSUE #5247: Reject deployments with empty error event, or ignore empty error events
**Describe the bug**

At the moment the user can deploy a process with an empty error event `<bpmn:error id="Error_1yxpabq" name="Error_2c7cj4n" />`. This will lead to the situation that the transformer fails to transform this diagram, but only when the diagram is loaded into the memory state.

**To Reproduce**

[empty-error.bpmn.txt](https://github.com/zeebe-io/zeebe/files/5134763/empty-error.bpmn.txt)

```
zbctl deploy --insecure empty-error.bpmn 
Error: rpc error: code = Unknown desc = Command rejected with code 'CREATE': Expected to create timer for start event, but encountered the following error: null

zbctl deploy --insecure empty-error.bpmn 
Error: rpc error: code = InvalidArgument desc = Command rejected with code 'CREATE': Expected to deploy new resources, but encountered the following errors:
'empty-error.bpmn': null
```

**Expected behavior**

Either reject the deployment or ignore empty error events.

**Log/Stacktrace**

<details><summary>Full Stacktrace</summary>
 <p>

```
2020-08-27 07:26:46.798 [Broker-0-StreamProcessor-1] [Broker-0-zb-actors-0] ERROR io.zeebe.broker.workflow - Unexpected error while processing resource 'empty-error.bpmn'         
java.lang.NullPointerException: null                                                                                                                                                                           at io.zeebe.util.StringUtil.getBytes(StringUtil.java:39) ~[zeebe-util-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]
        at io.zeebe.util.StringUtil.getBytes(StringUtil.java:35) ~[zeebe-util-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]                            
        at io.zeebe.util.buffer.BufferUtil.wrapString(BufferUtil.java:42) ~[zeebe-util-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]                                                                    
        at io.zeebe.engine.processing.deployment.model.transformer.ErrorTransformer.transform(ErrorTransformer.java:28) ~[zeebe-workflow-engine-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]   
        at io.zeebe.engine.processing.deployment.model.transformer.ErrorTransformer.transform(ErrorTransformer.java:17) ~[zeebe-workflow-engine-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]         
        at io.zeebe.engine.processing.deployment.model.transformation.TransformationVisitor.visit(TransformationVisitor.java:42) ~[zeebe-workflow-engine-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]
        at io.zeebe.model.bpmn.traversal.TypeHierarchyVisitor.visit(TypeHierarchyVisitor.java:39) ~[zeebe-bpmn-model-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]
        at io.zeebe.model.bpmn.traversal.ModelWalker.walk(ModelWalker.java:60) ~[zeebe-bpmn-model-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]
        at io.zeebe.engine.processing.deployment.model.transformation.BpmnTransformer.transformDefinitions(BpmnTransformer.java:97) ~[zeebe-workflow-engine-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]
        at io.zeebe.engine.state.deployment.WorkflowPersistenceCache.updateInMemoryState(WorkflowPersistenceCache.java:147) ~[zeebe-workflow-engine-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]
```
</p>
</details>

**Environment:**
 - OS: Docker
 - Zeebe Version: SNAPSHOT
 - Configuration: default

### LINKED ISSUE #5248: Reject Deployments with invalid timer cycles
**Describe the bug**

If a user deploys a process with an invalid timer cycle for example on a boundary event, like `"10"` the process is not rejected and on creating a new instance the process ends in an incident.

**To Reproduce**

Deploy and create an instance of the process [invalid-timer-cycle.bpmn.txt](https://github.com/zeebe-io/zeebe/files/5134793/invalid-timer-cycle.bpmn.txt)

**Expected behavior**

The process should be rejected on deployment.

**Log/Stacktrace**

<details><summary>Full Stacktrace</summary>
 <p>

```
2020-08-27 07:47:59.223 [Broker-0-Exporter-1] [Broker-0-zb-fs-workers-0] DEBUG io.zeebe.broker.exporter.DebugLogExporter - {"partitionId":1,"value":{"version":1,"flowScopeKey":2251799813685257,"bpmnElementType":"SERVICE_TASK","parentWorkflowInstanceKey":-1,"parentElementInstanceKey":-1,"elementId":"Activity_0ou5gje","workflowKey":2251799813685249,"workflowInstanceKey":2251799813685257,"bpmnProcessId":"invalid-timer-cycle"},"sourceRecordPosition":27,"position":28,"timestamp":1598514479191,"valueType":"WORKFLOW_INSTANCE","recordType":"EVENT","intent":"ELEMENT_ACTIVATING","rejectionType":"NULL_VAL","rejectionReason":"","key":2251799813685261}
2020-08-27 07:47:59.235 [Broker-0-Exporter-1] [Broker-0-zb-fs-workers-0] DEBUG io.zeebe.broker.exporter.DebugLogExporter - {"partitionId":1,"value":{"errorMessage":"Repetition spec must start with R","elementId":"Activity_0ou5gje","elementInstanceKey":2251799813685261,"workflowKey":2251799813685249,"workflowInstanceKey":2251799813685257,"bpmnProcessId":"invalid-timer-cycle","errorType":"EXTRACT_VALUE_ERROR","jobKey":-1,"variableScopeKey":2251799813685261},"sourceRecordPosition":28,"position":29,"timestamp":1598514479209,"valueType":"INCIDENT","recordType":"COMMAND","intent":"CREATE","rejectionType":"NULL_VAL","rejectionReason":"","key":-1}
2020-08-27 07:47:59.242 [Broker-0-Exporter-1] [Broker-0-zb-fs-workers-0] DEBUG io.zeebe.broker.exporter.DebugLogExporter - {"partitionId":1,"value":{"errorMessage":"Repetition spec must start with R","elementId":"Activity_0ou5gje","elementInstanceKey":2251799813685261,"workflowKey":2251799813685249,"workflowInstanceKey":2251799813685257,"bpmnProcessId":"invalid-timer-cycle","errorType":"EXTRACT_VALUE_ERROR","jobKey":-1,"variableScopeKey":2251799813685261},"sourceRecordPosition":29,"position":30,"timestamp":1598514479229,"valueType":"INCIDENT","recordType":"EVENT","intent":"CREATED","rejectionType":"NULL_VAL","rejectionReason":"","key":2251799813685262}
```

</p>
</details>

**Environment:**
 - OS: docker
 - Zeebe Version: SNAPSHOT
 - Configuration: default

### LINKED ISSUE #5253: Reject deployments with null output mapping
**Describe the bug**

If the output mapping target is `null` the deployment is corrupt but still is added to the state. Following deployments of the process are failing to parse the previous version of the process.

**To Reproduce**

Deploy [null-output-target.bpmn.txt](https://github.com/zeebe-io/zeebe/files/5140759/null-output-target.bpmn.txt)

**Expected behavior**

The deployment is rejected.

**Log/Stacktrace**

<details><summary>Full Stacktrace</summary>
 <p>

```
2020-08-28 07:47:26.970 [Broker-0-StreamProcessor-1] [Broker-0-zb-actors-1] ERROR io.zeebe.broker.workflow - Unexpected error while processing resource '/home/menski/null-output-target.bpmn'         
java.lang.IllegalStateException: Failed to build variable mapping expression: failed to parse expression '{outputTest1:"outputTest1",null:"outputTest2"}': [1.27] failure: '}' expected but ',' found  
                                                                                                                                                                                                       
{outputTest1:"outputTest1",null:"outputTest2"}                                                                                                                                                         
                          ^                                                                                                                                                                            
        at io.zeebe.engine.processing.deployment.model.transformer.VariableMappingTransformer.parseExpression(VariableMappingTransformer.java:180) ~[zeebe-workflow-engine-0.25.0-SNAPSHOT.jar:0.25.0-S
NAPSHOT]                                                                                                                                                                                               
        at io.zeebe.engine.processing.deployment.model.transformer.VariableMappingTransformer.transformOutputMappings(VariableMappingTransformer.java:88) ~[zeebe-workflow-engine-0.25.0-SNAPSHOT.jar:0
.25.0-SNAPSHOT]                                                                                                                                                                                        
        at io.zeebe.engine.processing.deployment.model.transformer.FlowNodeTransformer.lambda$transformIoMappings$3(FlowNodeTransformer.java:74) ~[zeebe-workflow-engine-0.25.0-SNAPSHOT.jar:0.25.0-SNA
PSHOT]                                                                                                                                                                                                 
        at java.util.Optional.map(Unknown Source) ~[?:?]                                                                                                                                               
        at io.zeebe.engine.processing.deployment.model.transformer.FlowNodeTransformer.transformIoMappings(FlowNodeTransformer.java:72) ~[zeebe-workflow-engine-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]
        at io.zeebe.engine.processing.deployment.model.transformer.FlowNodeTransformer.transform(FlowNodeTransformer.java:38) ~[zeebe-workflow-engine-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]             
        at io.zeebe.engine.processing.deployment.model.transformer.FlowNodeTransformer.transform(FlowNodeTransformer.java:21) ~[zeebe-workflow-engine-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]             
        at io.zeebe.engine.processing.deployment.model.transformation.TransformationVisitor.visit(TransformationVisitor.java:42) ~[zeebe-workflow-engine-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]          
        at io.zeebe.model.bpmn.traversal.TypeHierarchyVisitor.visit(TypeHierarchyVisitor.java:39) ~[zeebe-bpmn-model-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]                                              
        at io.zeebe.model.bpmn.traversal.ModelWalker.walk(ModelWalker.java:60) ~[zeebe-bpmn-model-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]                                                                 
        at io.zeebe.engine.processing.deployment.model.transformation.BpmnTransformer.transformDefinitions(BpmnTransformer.java:100) ~[zeebe-workflow-engine-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]
        at io.zeebe.engine.state.deployment.WorkflowPersistenceCache.updateInMemoryState(WorkflowPersistenceCache.java:147) ~[zeebe-workflow-engine-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]
        at io.zeebe.engine.state.deployment.WorkflowPersistenceCache.lookupWorkflowByIdAndPersistedVersion(WorkflowPersistenceCache.java:213) ~[zeebe-workflow-engine-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSH
OT]
        at io.zeebe.engine.state.deployment.WorkflowPersistenceCache.getLatestWorkflowVersionByProcessId(WorkflowPersistenceCache.java:194) ~[zeebe-workflow-engine-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT
]
        at io.zeebe.engine.state.deployment.WorkflowState.getLatestWorkflowVersionByProcessId(WorkflowState.java:62) ~[zeebe-workflow-engine-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]
        at io.zeebe.engine.processing.deployment.transform.DeploymentTransformer.transformWorkflowResource(DeploymentTransformer.java:154) ~[zeebe-workflow-engine-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]
        at io.zeebe.engine.processing.deployment.transform.DeploymentTransformer.transformResource(DeploymentTransformer.java:110) ~[zeebe-workflow-engine-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]
        at io.zeebe.engine.processing.deployment.transform.DeploymentTransformer.transform(DeploymentTransformer.java:81) ~[zeebe-workflow-engine-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]
        at io.zeebe.engine.processing.deployment.TransformingDeploymentCreateProcessor.processRecord(TransformingDeploymentCreateProcessor.java:71) ~[zeebe-workflow-engine-0.25.0-SNAPSHOT.jar:0.25.0-
SNAPSHOT]
        at io.zeebe.engine.processing.streamprocessor.TypedRecordProcessor.processRecord(TypedRecordProcessor.java:51) ~[zeebe-workflow-engine-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]
        at io.zeebe.engine.processing.streamprocessor.ProcessingStateMachine.lambda$processInTransaction$2(ProcessingStateMachine.java:267) ~[zeebe-workflow-engine-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT
]
        at io.zeebe.db.impl.rocksdb.transaction.ZeebeTransaction.run(ZeebeTransaction.java:79) ~[zeebe-db-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]
        at io.zeebe.engine.processing.streamprocessor.ProcessingStateMachine.processInTransaction(ProcessingStateMachine.java:258) ~[zeebe-workflow-engine-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]
        at io.zeebe.engine.processing.streamprocessor.ProcessingStateMachine.processEvent(ProcessingStateMachine.java:227) ~[zeebe-workflow-engine-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]
        at io.zeebe.engine.processing.streamprocessor.ProcessingStateMachine.tryToReadNextEvent(ProcessingStateMachine.java:203) ~[zeebe-workflow-engine-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]
        at io.zeebe.engine.processing.streamprocessor.ProcessingStateMachine.readNextEvent(ProcessingStateMachine.java:194) ~[zeebe-workflow-engine-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]
        at io.zeebe.util.sched.ActorJob.invoke(ActorJob.java:76) [zeebe-util-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]
        at io.zeebe.util.sched.ActorJob.execute(ActorJob.java:39) [zeebe-util-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]
        at io.zeebe.util.sched.ActorTask.execute(ActorTask.java:118) [zeebe-util-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]
        at io.zeebe.util.sched.ActorThread.executeCurrentTask(ActorThread.java:107) [zeebe-util-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]
        at io.zeebe.util.sched.ActorThread.doWork(ActorThread.java:91) [zeebe-util-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]
        at io.zeebe.util.sched.ActorThread.run(ActorThread.java:204) [zeebe-util-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]
```

</p>
</details>


**Environment:**
 - OS: Docker
 - Zeebe Version: SNAPSHOT
 - Configuration: default

### PR DESCRIPTION
**Pull request by saig0**
## Description

* validate the target expressions of variable mappings
* validate static timer expressions
* ignore error events that are not referenced in the workflow on validation and transformation
* transform the workflow on validation to avoid missing validations, or unexpected failures on transforming

## Related issues

closes #5253 
closes #5248 
closes #5247 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...develop?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The impact of the changes is verified by a benchmark 

Documentation: 
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)

### CODE REVIEW

**Comment by saig0** [comment]
@korthout I recommend reviewing the changes commit-by-commit. In the second commit, I migrated one test from the first commit to JUnit 5. I like to get your input for this new approach. Maybe, we can have a chat directly on this part.

**Review by korthout — COMMENTED** [review]
I only have 1 question on the first commit, but let's discuss the junit5 changes f2f

File: engine/src/main/java/io/zeebe/engine/processing/common/ExpressionProcessor.java [hunk 1]
```diff
@@ -130,10 +130,7 @@ public ExpressionProcessor(
           return Either.right(Interval.parse(result.getString()));
         } catch (final DateTimeParseException e) {
           return Either.left(
-              new Failure(
-                  String.format(
-                      "Expected result of the expression '%s' to be parsed to a duration, but was '%s' and encountered error: %s",
-                      expression.getExpression(), result.getString(), e.getMessage())));
+              new Failure(String.format("Invalid duration format '%s'", result.getString())));
```
[the remark this answers is not in the document]
Comment (korthout) [reply engine/src/main/java/io/zeebe/engine/processing/common/ExpressionProcessor.java:133]: This comment goes for all error messages changed in this PR

[the remark this answers is not in the document]
Comment (saig0) [reply engine/src/main/java/io/zeebe/engine/processing/common/ExpressionProcessor.java:133]: The message template ("Expected ... but was ...") is applied to log statements or exceptions. Here, the message is used for the validation (i.e. the deployment rejection message) and for creating an incident. So, I think it is okay to have a different message.

In these two cases, the message feels a bit long and it doesn't provide more useful information. 

`"Expected result of the expression '"foo"' to be parsed to a duration, but was '"foo"' and encountered error: Text cannot be parsed to a Period"`

vs.

`"Invalid duration format '"foo"'"`

[the remark this answers is not in the document]
Comment (saig0) [reply engine/src/main/java/io/zeebe/engine/processing/common/ExpressionProcessor.java:133]: Or, let's add the expression to the failure message.

**Review by korthout — APPROVED** [review]
As discussed, the junit5 changes look good. So, after you've pushed your last changes feel free to merge.

**Pull request merged**